### PR TITLE
Fixing error message with ReuseSourceFolderStructure restore switch 

### LINKED ISSUE #1799: Copy-DbaDatabase ReuseFolderStructure fails on 0.9.7 build
## Is this:
_Just place "x" to indicate selection (e.g. "[x]")_

- [ ] *feature* 
- [ x] *bug*:

## System Details

- Operating system name and version:
- Output from `$PSVersionTable`:

```
Name                           Value                                                                                                                                                                                                 
----                           -----                                                                                                                                                                                                 
PSVersion                      4.0                                                                                                                                                                                                   
WSManStackVersion              3.0                                                                                                                                                                                                   
SerializationVersion           1.1.0.1                                                                                                                                                                                               
CLRVersion                     4.0.30319.42000                                                                                                                                                                                       
BuildVersion                   6.3.9600.17400                                                                                                                                                                                        
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0}                                                                                                                                                                                  
PSRemotingProtocolVersion      2.2         
```                                                                                                                                                                                          

Evaluate $PSVersionTable in PowerShell and paste the output here
```
```
Name                           Value                                                                                                                                                                                                 
----                           -----                                                                                                                                                                                                 
PSVersion                      4.0                                                                                                                                                                                                   
WSManStackVersion              3.0                                                                                                                                                                                                   
SerializationVersion           1.1.0.1                                                                                                                                                                                               
CLRVersion                     4.0.30319.42000                                                                                                                                                                                       
BuildVersion                   6.3.9600.17400                                                                                                                                                                                        
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0}                                                                                                                                                                                  
PSRemotingProtocolVersion      2.2       
```                                                                                                                                                                                            

- Output of dbatools version:

```
Evaluate (Get-Module dbatools -ListAvailable).Version and paste output here
```
```
Major  Minor  Build  Revision
-----  -----  -----  --------
0      9      7      -1      

```
- SQL Server version for source/target

**Source** - Microsoft SQL Server 2008 (SP4) - 10.0.6000.29 (X64) 
	Sep  3 2014 04:11:34 
	Copyright (c) 1988-2008 Microsoft Corporation
	Enterprise Edition (64-bit) on Windows NT 6.1 <X64> (Build 7601: Service Pack 1)

**Destination** - Microsoft SQL Server 2014 (SP2) (KB3171021) - 12.0.5000.0 (X64) 
	Jun 17 2016 19:14:09 
	Copyright (c) Microsoft Corporation
	Enterprise Edition (64-bit) on Windows NT 6.3 <X64> (Build 9600: )



## Steps to Reproduce

#### Example:
1. Created database/object/files/etc
2. Run Command:
```powershell
Get-DbaFunction -SqlInstance MyServer | Get-DbaSomeOtherFunction -FilterSwitch -StelleEyedMissileMan
```
3. Received following error
```https://gist.github.com/TheRockStarDBA/765fe010cd12fc8aa47b095373aa7f27

Also

7/20/2017 5:02:51 PM

At line:582 char:38
+                     $FilteredFiles | Restore-DBFromFilteredArray -SqlInstance $S ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
System.Management.Automation.ParameterBindingValidationException: Cannot bind argument to parameter 'Message' because it is an empty string.
   at System.Management.Automation.ExceptionHandlingOps.CheckActionPreference(FunctionContext funcContext, Exception exception)
   at System.Management.Automation.Interpreter.ActionCallInstruction`2.Run(InterpretedFrame frame)
   at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
   at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
   at System.Management.Automation.Interpreter.Interpreter.Run(InterpretedFrame frame)
   at System.Management.Automation.Interpreter.LightLambda.RunVoid1[T0](T0 arg0)
   at System.Management.Automation.PSScriptCmdlet.RunClause(Action`1 clause, Object dollarUnderbar, Object inputToProcess)
   at System.Management.Automation.PSScriptCmdlet.DoEndProcessing()
   at System.Management.Automation.CommandProcessorBase.Complete()
7/20/2017 5:02:52 PM
```

### PR DESCRIPTION
**Pull request by Stuart-Moore**
## Type of Change

 - [x] Bug fix (non-breaking change, fixes #1799 and #1801  )
 - [ ] New feature (non-breaking change, adds functionality)
 - [ ] Breaking change (effects multiple commands or functionality)
 - [ ] Ran manual Pester test and has passed (`.\tests\manual.pester.ps1)
 - [ ] Pester test is included

### Purpose
Improves the handling of testing to see if we can reuse the original file path of the files being restored.

And surfaces the fail in a friendlier way

### Approach
Better logic to check for file existance, then folder existance and finally creating the folder if not there already

### Screenshots

image

Before, and after

### Learning
Use a bit of paper to draw out the logic tree first

### CODE REVIEW

**Comment by Stuart-Moore** [comment]
@FriedrichWeinmann - asked for you to review as I was having trouble surfacing the error from Restore-DbFromFilteredArray L167 to the catch in Restore-DbaDatabase on L585

Current implementation works, but feels a bit of a hack. How should I be doing this?

**Comment by potatoqualitee** [comment]
Will wait for @FriedrichWeinmann to merge 👍

File: internal/Restore-DBFromFilteredArray.ps1 [hunk 1]
```diff
@@ -141,22 +141,33 @@ Function Restore-DBFromFilteredArray {
...
                     }
+
                 }
                 else {
-                    Write-Message -Level Verbose -Message "Destination File $File  exists on $SqlInstance"
-                }
+                    Write-Message -Level Veryverbose -Message "Bombing out created on $sqlinstance"
+                    #Stop-Function -message "Destination File $File  exists on $SqlInstance" -Target $file -Category 'DeviceError' -silent $true
+                    Throw "Destination File $File  exists on $SqlInstance"
+                    return
```
Comment (FriedrichWeinmann) [inline internal/Restore-DBFromFilteredArray.ps1]: Use this instead:
```PowerShell
Stop-Function -message "Destination File $File  exists on $SqlInstance" -Target $file -Category 'DeviceError' -silent $true
return
```
(See above on how to catch this)

File: functions/Restore-DbaDatabase.ps1 [hunk 2]
```diff
@@ -583,7 +583,12 @@ function Restore-DbaDatabase {
                     $Completed = 'successfully'
                 }
                 catch {
-                    Write-Exception $_
+                    if ($_.FullyQualifiedErrorId -like 'Destination File*' ) {
+                        Stop-Function -Message "Restore of $databasename failed, $($_.FullyQualifiedErrorId) " -InnerErrorRecord $_.Exception.InnerException -silent:$silent
```
Comment (FriedrichWeinmann) [inline functions/Restore-DbaDatabase.ps1]: Catch like this instead:
```PowerShell
if ($_.CategoryInfo.Category -like "DeviceError") {
    Stop-Function -Message "Restore of $databasename failed, $_" -ErrorRecord $_ -Silent $silent
}
```
This allows you to reuse the original message same as before and you avoid parsing text to decide (which is what error categories are for).
See below on the proposed changes in `Restore-DBFromFilteredArray`

File: functions/Restore-DbaDatabase.ps1 [hunk 3]
```diff
@@ -583,7 +583,12 @@ function Restore-DbaDatabase {
                     $Completed = 'successfully'
                 }
                 catch {
-                    Write-Exception $_
+                    if ($_.FullyQualifiedErrorId -like 'Destination File*' ) {
+                        Stop-Function -Message "Restore of $databasename failed, $($_.FullyQualifiedErrorId) " -InnerErrorRecord $_.Exception.InnerException -silent:$silent    
+                    }
+                    else {
+                        Stop-Function -Message "Restore of $databasename failed" -InnerErrorRecord $_.Exception.InnerException -silent:$silent
```
Comment (FriedrichWeinmann) [inline functions/Restore-DbaDatabase.ps1]: Should instead be:
`Stop-Function -Message "Restore of $databasename failed" -ErrorRecord $_ -Silent $silent`

File: functions/Restore-DbaDatabase.ps1 [hunk 4]
```diff
@@ -557,7 +557,7 @@ function Restore-DbaDatabase {
 		
 		
         if ($AllFilteredFiles.count -gt 1 -and $DatabaseName -ne '') {
-            Write-Warning "$FunctionName -  DatabaseName parameter and multiple database restores is not compatible "
+            Write-Message -Level Warning -Message "$FunctionName -  DatabaseName parameter and multiple database restores is not compatible " 
             break
```
Comment (FriedrichWeinmann) [inline functions/Restore-DbaDatabase.ps1]: Use Stop-Function instead. Including the function name in messages is no longer needed.

File: functions/Restore-DbaDatabase.ps1 [hunk 5]
```diff
@@ -533,7 +533,7 @@ function Restore-DbaDatabase {
             if ($DestinationLogDirectory -ne '') {
                 if ((Test-DbaSqlPath -Path $DestinationLogDirectory -SqlInstance $SqlInstance -SqlCredential $SqlCredential) -ne $true) {
                     if ((New-DbaSqlDirectory -Path $DestinationLogDirectory -SqlInstance $SqlInstance -SqlCredential $SqlCredential).Created -ne $true) {
-                        Write-Warning "$FunctionName - DestinationLogDirectory $DestinationLogDirectory does not exist, and could not be created on $SqlInstance"
+                        Write-Message -Level Warning -Message "$FunctionName - DestinationLogDirectory $DestinationLogDirectory does not exist, and could not be created on $SqlInstance" 
                         break
```
Comment (FriedrichWeinmann) [inline functions/Restore-DbaDatabase.ps1]: Use Stop-Function instead. Including the function name in messages is no longer needed.

File: functions/Restore-DbaDatabase.ps1 [hunk 6]
```diff
@@ -519,7 +519,7 @@ function Restore-DbaDatabase {
             if ($DestinationDataDirectory -ne '') {
                 if ((Test-DbaSqlPath -Path $DestinationDataDirectory -SqlInstance $SqlInstance -SqlCredential $SqlCredential) -ne $true) {
                     if ((New-DbaSqlDirectory -Path $DestinationDataDirectory -SqlInstance $SqlInstance -SqlCredential $SqlCredential).Created -ne $true) {
-                        Write-Warning "$FunctionName - DestinationDataDirectory $DestinationDataDirectory does not exist, and could not be created on $SqlInstance"
+                        Write-Message -Level Warning -Message "$FunctionName - DestinationDataDirectory $DestinationDataDirectory does not exist, and could not be created on $SqlInstance" 
                         break
```
Comment (FriedrichWeinmann) [inline functions/Restore-DbaDatabase.ps1]: Use Stop-Function instead. Including the function name in messages is no longer needed.

File: functions/Restore-DbaDatabase.ps1 [hunk 7]
```diff
@@ -509,7 +509,7 @@ function Restore-DbaDatabase {
 		
         if ($null -ne $DatabaseName) {
             If (($DatabaseName -in ($server.Databases.name)) -and ($WithReplace -eq $false)) {
-                Write-Warning "$FunctionName - $DatabaseName exists on Sql Instance $SqlInstance , must specify WithReplace to continue"
+                Write-Message -Level Warning -Message "$FunctionName - $DatabaseName exists on Sql Instance $SqlInstance , must specify WithReplace to continue" 
                 break
```
Comment (FriedrichWeinmann) [inline functions/Restore-DbaDatabase.ps1]: Use Stop-Function instead. Including the function name in messages is no longer needed.

File: functions/Restore-DbaDatabase.ps1 [hunk 8]
```diff
@@ -469,7 +469,7 @@ function Restore-DbaDatabase {
                             if ($FileTmp.PSobject.Properties.name -match "Server") {
                                 #Most likely incoming from Get-DbaBackupHistory
                                 if ($Filetmp.Server -ne $SqlInstance -and $FileTmp.FullName -notlike '\\*') {
-                                    Write-Warning "$FunctionName - Backups from a different server and on a local drive, can't access"
+                                    Write-Message -Level Warning -Message "$FunctionName - Backups from a different server and on a local drive, can't access" 
                                     return
```
Comment (FriedrichWeinmann) [inline functions/Restore-DbaDatabase.ps1]: Use Stop-Function instead.

File: functions/Restore-DbaDatabase.ps1 [hunk 9]
```diff
@@ -426,7 +426,7 @@ function Restore-DbaDatabase {
                                     $backupFiles += $p
                                 }
                                 else {
-                                    Write-Warning "$FunctionName - $p cannot be accessed by $SqlInstance"
+                                    Write-Message -Level Warning -Message "$FunctionName - $p cannot be accessed by $SqlInstance" 
                                     continue
```
Comment (FriedrichWeinmann) [inline functions/Restore-DbaDatabase.ps1]: Stop-Function is designed to handle this.
 - Use `-Continue` parameter to process next item
 - Use `-SilentlyContinue` parameter if you want it to continue even in unattended mode (e.g. with scripts calling it. They won't have a way to know this failed if you do.

File: internal/Restore-DBFromFilteredArray.ps1 [hunk 10]
```diff
@@ -351,7 +362,7 @@ Function Restore-DBFromFilteredArray {
                     Write-Message -Level Verbose -Message "Failed, Closing Server connection"
                     $RestoreComplete = $False
                     $ExitError = $_.Exception.InnerException
-                    Write-Warning "$FunctionName - $ExitError" -WarningAction stop
+                    Stop-Function -Message  "Failed to restore db $DbName, stopping" -InnerErrorRecord $ExitError 
                     #Exit as once one restore has failed there's no point continuing
                     break
```
Comment (FriedrichWeinmann) [inline internal/Restore-DBFromFilteredArray.ps1]: Can't use it like this: Stop-Function does not accept exceptions, just error records (hence the name)
Also please avoid `break` if it isn't necessary (even though it doesn't do harm here)
Options:
```PowerShell
Stop-Function -Message  "Failed to restore db $DbName, stopping" -ErrorRecord $_
return
```
or
```PowerShell
$record = New-Object System.Management.Automation.ErrorRecord($_.Exception.InnerException, "manual", "InvalidOperation", $DbName)
Stop-Function -Message  "Failed to restore db $DbName, stopping" -ErrorRecord $record
return
```
If you must have the inner exception as the official record.

**Comment by potatoqualitee** [comment]
Hello,
I made all the requested changes and all tests have passed. Please review and merge at your convenience.

**Comment by potatoqualitee** [comment]
teehee

**Pull request merged**